### PR TITLE
Refactor auth

### PR DESCRIPTION
**Pull request by jbrunton**

### CODE REVIEW

**Comment by jbrunton-ci-minion** [comment]
#### :tropical_drink: `up` on refactor-auth

   <details>
   <summary>Click to expand Pulumi report</summary>

   ```
   Updating (refactor-auth)

View Live: https://app.pulumi.com/jbrunton/chat-demo/refactor-auth/updates/6


   pulumi:pulumi:Stack chat-demo-refactor-auth running 
   pulumi:pulumi:Stack chat-demo-refactor-auth running stack config: {
~  digitalocean:index:App chat-demo-refactor-auth updating [diff: ~spec]
@ Updating......
~  digitalocean:index:App chat-demo-refactor-auth updated [diff: ~spec]
   pulumi:pulumi:Stack chat-demo-refactor-auth  1 message

Diagnostics:
 pulumi:pulumi:Stack (chat-demo-refactor-auth):
   stack config: {
    "stackName": "refactor-auth",
    "appName": "chat-demo-refactor-auth",
    "environment": "development",
    "tag": "HEAD-110f89c",
    "protect": false,
    "domain": "refactor-auth.chat-demo.dev.jbrunton-do.com",
    "publicUrl": "https://refactor-auth.chat-demo.dev.jbrunton-do.com",
    "rootDomain": "jbrunton-do.com",
    "specId": "chat-demo-refactor-auth/HEAD-110f89c/fce93ab8"
   }

Outputs:
   publicUrl: "https://refactor-auth.chat-demo.dev.jbrunton-do.com"

Resources:
   ~ 1 updated
   1 unchanged

Duration: 1m18s


   ```
   
   </details>

File: server/src/app/auth/authenticate.spec.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  const testAdapterUser: AdapterUser = {
+    ...testUser,
+    emailVerified: new Date(),
+  };
+
+  const testSession: Session = {
+    user: testUser,
+    expires: new Date().toISOString(),
+  };
```
Comment (jbrunton) [inline server/src/app/auth/authenticate.spec.ts:30]: Create some fixtures for this, e.g.

```
const { user } = mockTestUser(dependencies, { createSession: true, createUser: true }); // options default to true
```

File: server/src/app/dependencies/index.ts [hunk 2]
```diff
@@ -45,6 +36,9 @@ export type AppDependencies = Dependencies & {
 };
 
 export type ReqDependencies = AppDependencies & {
+  req: NextApiRequest;
+  res: NextApiResponse;
```
Comment (jbrunton) [inline server/src/app/dependencies/index.ts]: Consider creating adapter types. Something like:

```
type ApiRequest<T> = {
    method: string;
    query: Record<string, string | string[] | undefined>;
    body: T;
    user?: user;
}

interface ApiResponse {
  sendResponse<T>(code: number, response: T);
}
```

File: server/src/app/messages/handle-message.spec.ts [hunk 3]
```diff
@@ -1,125 +1,100 @@
...
-describe("parseMessage", () => {
-  const time = "2022-01-01T10:30:00.000Z";
-  const roomId = "a1b2c3";
-  const sender = {
-    id: "1234",
-    name: "User_123",
-  };
+describe("handleMessage", () => {
+  let dispatcher: MockProxy<Dispatcher>;
+  const now = new Date("2022-02-02T21:22:23.234Z");
```
Comment (jbrunton) [inline server/src/app/messages/handle-message.spec.ts:16]: is this used?

File: server/src/app/messages/handle-message.spec.ts [hunk 4]
```diff
@@ -1,125 +1,100 @@
...
-    deps = {
-      userRepository,
-      roomRepository,
-      auditLogRepository: mock<AuditLogRepository>(),
-      nameGenerator: mock<NameGenerator>(),
-      dispatcher,
-      mailer: mock<Mailer>(),
-      adapter: mock<Adapter>(),
-    };
+    deps = mockReqDependencies();
```
Comment (jbrunton) [inline server/src/app/messages/handle-message.spec.ts:53]: Provide overridable params, e.g.

```
mockReqDependencies({
  req: {
    method: "POST",
    query: { id: roomId }
  }
})
```

File: server/src/app/messages/handle-message.ts [hunk 5]
```diff
@@ -1,59 +1,74 @@
...
+    RT.chain(processCommands),
+    RT.chain(processResponse),
+    RT.chain(sendResponse)
+  );
 };
 
-export const handleMessage = (
+const validateRoom = (
+  message: PublicMessage | Command
+): DependencyReaderTask<PublicMessage | Command, ReqDependencies> =>
```
Comment (jbrunton) [inline server/src/app/messages/handle-message.ts:34]: Check if the specific dependencies type (on this and below functions) can be narrowed.

**Pull request merged**